File: sdk-swift/Sources/DecibelSDK/Client/DecibelReadClient.swift
import Foundation

/// DecibelReadClient provides read-only access to the Decibel API.
public class DecibelReadClient {
    private let config: DecibelConfig
    private let apiKey: String?
    private let session: URLSession
    private let decoder: JSONDecoder

    public init(config: DecibelConfig, apiKey: String? = nil) {
        self.config = config
        self.apiKey = apiKey
        self.session = URLSession.shared

        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        self.decoder = decoder
    }

    public func getAllMarkets() async throws -> [PerpMarketConfig] {
        return try await performRequest(endpoint: "/markets")
    }

    public func getMarketByName(name: String) async throws -> PerpMarketConfig {
        return try await performRequest(endpoint: "/markets/\(name)")
    }

    public func getMarketDepth(marketName: String, limit: Int? = nil) async throws -> MarketDepth {
        var endpoint = "/market_depth/\(marketName)"
        if let limit = limit {
            endpoint += "?limit=\(limit)"
        }
        return try await performRequest(endpoint: endpoint)
    }

    public func getAllMarketPrices() async throws -> [MarketPrice] {
        return try await performRequest(endpoint: "/market_prices")
    }

    public func getMarketPriceByName(name: String) async throws -> [MarketPrice] {
        return try await performRequest(endpoint: "/market_prices/\(name)")
    }

    public func getMarketTrades(marketName: String, limit: Int? = nil) async throws -> [MarketTrade] {
        var endpoint = "/market_trades/\(marketName)"
        if let limit = limit {
            endpoint += "?limit=\(limit)"
        }
        return try await performRequest(endpoint: endpoint)
    }

    public func getCandlesticks(
        marketName: String,
        interval: String,
        startTime: Int64? = nil,
        endTime: Int64? = nil
    ) async throws -> [Candlestick] {
        var endpoint = "/candlesticks/\(marketName)/\(interval)"
        var queryParams = [String]()
        if let startTime = startTime {
            queryParams.append("start_time=\(startTime)")
        }
        if let endTime = endTime {
            queryParams.append("end_time=\(endTime)")
        }
        if !queryParams.isEmpty {
            endpoint += "?" + queryParams.joined(separator: "&")
        }
        return try await performRequest(endpoint: endpoint)
    }

    public func getAllMarketContexts() async throws -> [MarketContext] {
        return try await performRequest(endpoint: "/market_contexts")
    }

    public func getAccountOverview(subaccountAddr: String) async throws -> AccountOverview {
        return try await performRequest(endpoint: "/account_overview/\(subaccountAddr)")
    }

    public func getUserPositions(subaccountAddr: String) async throws -> [UserPosition] {
        return try await performRequest(endpoint: "/positions/\(subaccountAddr)")
    }

    public func getUserOpenOrders(subaccountAddr: String) async throws -> [UserOpenOrder] {
        return try await performRequest(endpoint: "/open_orders/\(subaccountAddr)")
    }

    public func getUserOrderHistory(
        subaccountAddr: String,
        limit: Int? = nil,
        offset: Int? = nil,
        marketName: String? = nil
    ) async throws -> [UserOrderHistoryItem] {
        var endpoint = "/order_history/\(subaccountAddr)"
        var queryParams = [String]()
        if let limit = limit { queryParams.append("limit=\(limit)") }
        if let offset = offset { queryParams.append("offset=\(offset)") }
        if let marketName = marketName { queryParams.append("market=\(marketName)") }
        if !queryParams.isEmpty {
            endpoint += "?" + queryParams.joined(separator: "&")
        }
        return try await performRequest(endpoint: endpoint)
    }

    public func getUserTradeHistory(
        subaccountAddr: String,
        limit: Int? = nil,
        offset: Int? = nil,
        marketName: String? = nil
    ) async throws -> [UserTradeHistoryItem] {
        var endpoint = "/trade_history/\(subaccountAddr)"
        var queryParams = [String]()
        if let limit = limit { queryParams.append("limit=\(limit)") }
        if let offset = offset { queryParams.append("offset=\(offset)") }
        if let marketName = marketName { queryParams.append("market=\(marketName)") }
        if !queryParams.isEmpty {
            endpoint += "?" + queryParams.joined(separator: "&")
        }
        return try await performRequest(endpoint: endpoint)
    }

    public func getUserFundingHistory(
        subaccountAddr: String,
        limit: Int? = nil,
        offset: Int? = nil,
        marketName: String? = nil
    ) async throws -> [UserFundingHistoryItem] {
        var endpoint = "/funding_history/\(subaccountAddr)"
        var queryParams = [String]()
        if let limit = limit { queryParams.append("limit=\(limit)") }
        if let offset = offset { queryParams.append("offset=\(offset)") }
        if let marketName = marketName { queryParams.append("market=\(marketName)") }
        if !queryParams.isEmpty {
            endpoint += "?" + queryParams.joined(separator: "&")
        }
        return try await performRequest(endpoint: endpoint)
    }

    public func getUserFundHistory(
        subaccountAddr: String,
        limit: Int? = nil,
        offset: Int? = nil
    ) async throws -> [UserFundHistoryItem] {
        var endpoint = "/fund_history/\(subaccountAddr)"
        var queryParams = [String]()
        if let limit = limit { queryParams.append("limit=\(limit)") }
        if let offset = offset { queryParams.append("offset=\(offset)") }
        if !queryParams.isEmpty {
            endpoint += "?" + queryParams.joined(separator: "&")
        }
        return try await performRequest(endpoint: endpoint)
    }

    public func getUserSubaccounts(ownerAddr: String) async throws -> [UserSubaccount] {
        return try await performRequest(endpoint: "/subaccounts/\(ownerAddr)")
    }

    public func getDelegations(subaccountAddr: String) async throws -> [Delegation] {
        return try await performRequest(endpoint: "/delegations/\(subaccountAddr)")
    }

    public func getActiveTwaps(subaccountAddr: String) async throws -> [UserActiveTwap] {
        return try await performRequest(endpoint: "/active_twaps/\(subaccountAddr)")
    }

    public func getTwapHistory(
        subaccountAddr: String,
        limit: Int? = nil,
        offset: Int? = nil
    ) async throws -> [UserTwapHistoryItem] {
        var endpoint = "/twap_history/\(subaccountAddr)"
        var queryParams = [String]()
        if let limit = limit { queryParams.append("limit=\(limit)") }
        if let offset = offset { queryParams.append("offset=\(offset)") }
        if !queryParams.isEmpty {
            endpoint += "?" + queryParams.joined(separator: "&")
        }
        return try await performRequest(endpoint: endpoint)
    }

    public func getVaults(limit: Int? = nil, offset: Int? = nil) async throws -> [Vault] {
        var endpoint = "/vaults"
        var queryParams = [String]()
        if let limit = limit { queryParams.append("limit=\(limit)") }
        if let offset = offset { queryParams.append("offset=\(offset)") }
        if !queryParams.isEmpty {
            endpoint += "?" + queryParams.joined(separator: "&")
        }
        return try await performRequest(endpoint: endpoint)
    }

    public func getUserOwnedVaults(ownerAddr: String) async throws -> [Vault] {
        return try await performRequest(endpoint: "/vaults/owner/\(ownerAddr)")
    }

    public func getUserPerformancesOnVaults(userAddr: String) async throws -> [VaultPerformance] {
        return try await performRequest(endpoint: "/vaults/performance/\(userAddr)")
    }

    public func getLeaderboard(limit: Int? = nil, offset: Int? = nil) async throws -> [LeaderboardEntry] {
        var endpoint = "/leaderboard"
        var queryParams = [String]()
        if let limit = limit { queryParams.append("limit=\(limit)") }
        if let offset = offset { queryParams.append("offset=\(offset)") }
        if !queryParams.isEmpty {
            endpoint += "?" + queryParams.joined(separator: "&")
        }
        return try await performRequest(endpoint: endpoint)
    }

    public func close() {}

    // MARK: - Private Methods

    private func performRequest<T: Decodable>(endpoint: String) async throws -> T {
        let urlString = config.tradingHTTPURL + endpoint
        guard let url = URL(string: urlString) else {
            throw DecibelError.config("Invalid URL: \(urlString)")
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if let apiKey = apiKey {
            request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        }

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw DecibelError.config("Invalid HTTP response")
        }

        guard httpResponse.statusCode < 400 else {
            let message = String(data: data, encoding: .utf8) ?? "Unknown error"
            throw DecibelError.api(
                status: httpResponse.statusCode,
                statusText: HTTPURLResponse.localizedString(forStatusCode: httpResponse.statusCode),
                message: message
            )
        }

        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            throw DecibelError.serialization("Failed to decode response: \(error.localizedDescription)")
        }
    }
}

File: sdk-kotlin/src/main/kotlin/trade/decibel/sdk/client/DecibelReadClient.kt
package trade.decibel.sdk.client

import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.request.*
import io.ktor.http.*
import trade.decibel.sdk.config.DecibelConfig
import trade.decibel.sdk.models.*

/**
 * DecibelReadClient provides read-only access to the Decibel API.
 */
class DecibelReadClient(
    private val config: DecibelConfig,
    private val apiKey: String? = null
) {
    private val client = HttpClient {
        expectSuccess = false
    }

    /**
     * Retrieves all market configurations.
     */
    suspend fun getAllMarkets(): List<PerpMarketConfig> {
        return client.get("${config.tradingHttpUrl}/markets") {
            setApiKey()
        }.body()
    }

    /**
     * Retrieves a specific market by name.
     */
    suspend fun getMarketByName(name: String): PerpMarketConfig {
        return client.get("${config.tradingHttpUrl}/markets/$name") {
            setApiKey()
        }.body()
    }

    /**
     * Retrieves the order book for a market.
     */
    suspend fun getMarketDepth(marketName: String, limit: Int? = null): MarketDepth {
        return client.get("${config.tradingHttpUrl}/market_depth/$marketName") {
            setApiKey()
            limit?.let { parameter("limit", it) }
        }.body()
    }

    /**
     * Retrieves prices for all markets.
     */
    suspend fun getAllMarketPrices(): List<MarketPrice> {
        return client.get("${config.tradingHttpUrl}/market_prices") {
            setApiKey()
        }.body()
    }

    /**
     * Retrieves the price for a specific market.
     */
    suspend fun getMarketPriceByName(name: String): List<MarketPrice> {
        return client.get("${config.tradingHttpUrl}/market_prices/$name") {
            setApiKey()
        }.body()
    }

    /**
     * Retrieves recent trades for a market.
     */
    suspend fun getMarketTrades(marketName: String, limit: Int? = null): List<MarketTrade> {
        return client.get("${config.tradingHttpUrl}/market_trades/$marketName") {
            setApiKey()
            limit?.let { parameter("limit", it) }
        }.body()
    }

    /**
     * Retrieves historical candlestick data.
     */
    suspend fun getCandlesticks(
        marketName: String,
        interval: String,
        startTime: Long? = null,
        endTime: Long? = null
    ): List<Candlestick> {
        return client.get("${config.tradingHttpUrl}/candlesticks/$marketName/$interval") {
            setApiKey()
            startTime?.let { parameter("start_time", it) }
            endTime?.let { parameter("end_time", it) }
        }.body()
    }

    /**
     * Retrieves context data for all markets.
     */
    suspend fun getAllMarketContexts(): List<MarketContext> {
        return client.get("${config.tradingHttpUrl}/market_contexts") {
            setApiKey()
        }.body()
    }

    /**
     * Retrieves the account overview for a subaccount.
     */
    suspend fun getAccountOverview(subaccountAddr: String): AccountOverview {
        return client.get("${config.tradingHttpUrl}/account_overview/$subaccountAddr") {
            setApiKey()
        }.body()
    }

    /**
     * Retrieves all positions for a subaccount.
     */
    suspend fun getUserPositions(subaccountAddr: String): List<UserPosition> {
        return client.get("${config.tradingHttpUrl}/positions/$subaccountAddr") {
            setApiKey()
        }.body()
    }

    /**
     * Retrieves all open orders for a subaccount.
     */
    suspend fun getUserOpenOrders(subaccountAddr: String): List<UserOpenOrder> {
        return client.get("${config.tradingHttpUrl}/open_orders/$subaccountAddr") {
            setApiKey()
        }.body()
    }

    /**
     * Retrieves the order history for a subaccount.
     */
    suspend fun getUserOrderHistory(
        subaccountAddr: String,
        limit: Int? = null,
        offset: Int? = null,
        marketName: String? = null
    ): List<UserOrderHistoryItem> {
        return client.get("${config.tradingHttpUrl}/order_history/$subaccountAddr") {
            setApiKey()
            limit?.let { parameter("limit", it) }
            offset?.let { parameter("offset", it) }
            marketName?.let { parameter("market", it) }
        }.body()
    }

    /**
     * Retrieves the trade history for a subaccount.
     */
    suspend fun getUserTradeHistory(
        subaccountAddr: String,
        limit: Int? = null,
        offset: Int? = null,
        marketName: String? = null
    ): List<UserTradeHistoryItem> {
        return client.get("${config.tradingHttpUrl}/trade_history/$subaccountAddr") {
            setApiKey()
            limit?.let { parameter("limit", it) }
            offset?.let { parameter("offset", it) }
            marketName?.let { parameter("market", it) }
        }.body()
    }

    /**
     * Retrieves the funding payment history for a subaccount.
     */
    suspend fun getUserFundingHistory(
        subaccountAddr: String,
        limit: Int? = null,
        offset: Int? = null,
        marketName: String? = null
    ): List<UserFundingHistoryItem> {
        return client.get("${config.tradingHttpUrl}/funding_history/$subaccountAddr") {
            setApiKey()
            limit?.let { parameter("limit", it) }
            offset?.let { parameter("offset", it) }
            marketName?.let { parameter("market", it) }
        }.body()
    }

    /**
     * Retrieves the deposit/withdrawal history for a subaccount.
     */
    suspend fun getUserFundHistory(
        subaccountAddr: String,
        limit: Int? = null,
        offset: Int? = null
    ): List<UserFundHistoryItem> {
        return client.get("${config.tradingHttpUrl}/fund_history/$subaccountAddr") {
            setApiKey()
            limit?.let { parameter("limit", it) }
            offset?.let { parameter("offset", it) }
        }.body()
    }

    /**
     * Retrieves all subaccounts for an owner address.
     */
    suspend fun getUserSubaccounts(ownerAddr: String): List<UserSubaccount> {
        return client.get("${config.tradingHttpUrl}/subaccounts/$ownerAddr") {
            setApiKey()
        }.body()
    }

    /**
     * Retrieves all delegations for a subaccount.
     */
    suspend fun getDelegations(subaccountAddr: String): List<Delegation> {
        return client.get("${config.tradingHttpUrl}/delegations/$subaccountAddr") {
            setApiKey()
        }.body()
    }

    /**
     * Retrieves all active TWAP orders for a subaccount.
     */
    suspend fun getActiveTwaps(subaccountAddr: String): List<UserActiveTwap> {
        return client.get("${config.tradingHttpUrl}/active_twaps/$subaccountAddr") {
            setApiKey()
        }.body()
    }

    /**
     * Retrieves the TWAP order history for a subaccount.
     */
    suspend fun getTwapHistory(
        subaccountAddr: String,
        limit: Int? = null,
        offset: Int? = null
    ): List<UserTwapHistoryItem> {
        return client.get("${config.tradingHttpUrl}/twap_history/$subaccountAddr") {
            setApiKey()
            limit?.let { parameter("limit", it) }
            offset?.let { parameter("offset", it) }
        }.body()
    }

    /**
     * Retrieves all vaults.
     */
    suspend fun getVaults(limit: Int? = null, offset: Int? = null): List<Vault> {
        return client.get("${config.tradingHttpUrl}/vaults") {
            setApiKey()
            limit?.let { parameter("limit", it) }
            offset?.let { parameter("offset", it) }
        }.body()
    }

    /**
     * Retrieves vaults owned by a user.
     */
    suspend fun getUserOwnedVaults(ownerAddr: String): List<Vault> {
        return client.get("${config.tradingHttpUrl}/vaults/owner/$ownerAddr") {
            setApiKey()
        }.body()
    }

    /**
     * Retrieves performance metrics for vaults the user has interacted with.
     */
    suspend fun getUserPerformancesOnVaults(userAddr: String): List<VaultPerformance> {
        return client.get("${config.tradingHttpUrl}/vaults/performance/$userAddr") {
            setApiKey()
        }.body()
    }

    /**
     * Retrieves the leaderboard.
     */
    suspend fun getLeaderboard(limit: Int? = null, offset: Int? = null): List<LeaderboardEntry> {
        return client.get("${config.tradingHttpUrl}/leaderboard") {
            setApiKey()
            limit?.let { parameter("limit", it) }
            offset?.let { parameter("offset", it) }
        }.body()
    }

    /**
     * Closes the HTTP client.
     */
    fun close() {
        client.close()
    }

    private fun HttpRequestBuilder.setApiKey() {
        apiKey?.let { header("x-api-key", it) }
        header(HttpHeaders.ContentType, ContentType.Application.Json)
    }
}
